### PR TITLE
[Sketcher] Add contextual BSpline commands in right click menu

### PR DESCRIPTION
**Pull request by maxwxyz**
Add B-Spline commands in the contextual right click menu in the Sketcher WB, depending of the selection of
- B-Splines
- Knots
- BSpline + Points

+fine tuned icons for better identification.
grafik
grafik

### CODE REVIEW

**Comment by maxwxyz** [comment]
noticed and adjusted
- tangent/perpendicular to line/point and point/point incl. point/bspline knot
- respect of axis and origin (previously no context menu when only axes or root point were selected)
- count axes as line and origin as point for the selection of possible constraints

**Review by AjinkyaDahale — COMMENTED** [review]
I haven't had the time to compile it myself but some minor questions that came up superficially.

File: src/Mod/Sketcher/Gui/ViewProviderSketch.cpp [hunk 1]
```diff
@@ -3824,31 +3828,69 @@ void ViewProviderSketch::generateContextMenu()
...
+                 << "Sketcher_BSplineDecreaseDegree";
+        }
+        else if (selectedBsplineKnots > 0 && selectedBsplineKnots == selectedPoints
+                 && selectedEdges == 0 && !onlyOrigin) {
+            if (selectedBsplineKnots == 1) {
+                menu << "Sketcher_BSplineIncreaseKnotMultiplicity"
+                     << "Sketcher_BSplineDecreaseKnotMultiplicity";
+            }
+            else if (selectedBsplineKnots == 2) {
+                menu << "Sketcher_JoinCurves";
```
Comment (AjinkyaDahale) [inline src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3890]: Join curves works for any two curves, not just bsplines. And only end points should be selected, not knots.

File: src/Mod/Sketcher/Gui/Resources/icons/splines/Sketcher_BSplineInsertKnot.svg
[the remark this answers is not in the document]
Comment (maxwxyz) [reply src/Mod/Sketcher/Gui/Resources/icons/splines/Sketcher_BSplineInsertKnot.svg:1]: it was green before, just the outline was added around the point.

File: src/Mod/Sketcher/Gui/ViewProviderSketch.cpp (hunk 1 above)
Comment (maxwxyz) [reply src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3890]: you are right, but it converts the "normal" geometry to splines. Not sure if we should add that in the context menu if not a spline end point is selected.

File: src/Mod/Sketcher/Gui/ViewProviderSketch.cpp [hunk 2]
```diff
@@ -3824,31 +3828,69 @@ void ViewProviderSketch::generateContextMenu()
     // if something is selected, count different elements in the current selection
     if (selection.size() > 0) {
         const std::vector<std::string> SubNames = selection[0].getSubNames();
+        const Sketcher::SketchObject* obj =
```
[the remark this answers is not in the document]
Comment (maxwxyz) [reply src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3831]: Is it possible to have a selection in an active sketch without an obj?

[the remark this answers is not in the document]
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3831]: Umm, maybe external selection? Or some badly designed tool?

File: src/Mod/Sketcher/Gui/ViewProviderSketch.cpp (hunk 1 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3890]: Why not? If we have selected two separate end points it does seem to make sense.

Comment (maxwxyz) [reply src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3890]: true. added it.

File: src/Mod/Sketcher/Gui/ViewProviderSketch.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxwxyz) [reply src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:3831]: added

**Comment by maxwxyz** [comment]
@AjinkyaDahale should be good now.

File: src/Mod/Sketcher/Gui/Resources/icons/splines/Sketcher_BSplineInsertKnot.svg
[the remark this answers is not in the document]
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/Gui/Resources/icons/splines/Sketcher_BSplineInsertKnot.svg:1]: @obelisk79 any opinion for this design choice?

**Comment by AjinkyaDahale** [comment]
Tried a few experiments. This seems to work. Is there anything specific I should try?

**Comment by maxwxyz** [comment]
@chennes is anything holding this back?

**Comment by chennes** [comment]
We are waiting for the Mac Conda CI to re-run, it had a glitch last time.

**Comment by maxwxyz** [comment]
Ah finally it ran through!

**Pull request merged**